File: wafel_py_old/ui/frame_slider.py
from typing import *

import wafel.imgui as ig
from wafel.util import *


def render_frame_slider(
  id: str,
  current_frame: int,
  num_frames: int,
  loaded_frames: List[int] = [],
) -> Maybe[int]:
  ig.push_id(id)

  pos = ig.get_cursor_pos()
  pos = (
    pos[0] + ig.get_window_position()[0],
    pos[1] + ig.get_window_position()[1] - ig.get_scroll_y(),
  )
  width = ig.get_content_region_available_width()

  ig.push_item_width(width)
  changed, new_frame = ig.slider_int(
    '##slider',
    current_frame,
    0,
    num_frames - 1,
  )
  ig.pop_item_width()

  dl = ig.get_window_draw_list()
  for frame in loaded_frames:
    line_pos = pos[0] + frame / num_frames * width
    dl.add_line(
      line_pos, pos[1] + 13,
      line_pos, pos[1] + 18,
      ig.get_color_u32_rgba(1, 0, 0, 1),
    )

  ig.pop_id()

  if changed:
    return Just(new_frame)
  else:
    return None


__all__ = ['render_frame_slider']

File: wafel_app/src/frame_slider.rs
use imgui::{self as ig, im_str};

// TODO: Hide loaded frames in release mode

pub(crate) fn render_frame_slider(
    ui: &ig::Ui<'_>,
    id: &str,
    current_frame: u32,
    max_frame: u32,
    loaded_frames: &[u32],
) -> Option<u32> {
    let id_token = ui.push_id(id);

    let mut pos = ui.cursor_pos();
    pos[0] += ui.window_pos()[0];
    pos[1] += ui.window_pos()[1] - ui.scroll_y();

    let width = ui.content_region_avail()[0];
    ui.set_next_item_width(width);

    let mut new_frame = current_frame;
    let changed = ig::Slider::new(im_str!("##slider"))
        .range(0..=max_frame)
        .build(ui, &mut new_frame);

    let dl = ui.get_window_draw_list();
    for &frame in loaded_frames {
        let line_pos = pos[0] + frame as f32 / (max_frame + 1) as f32 * width;
        dl.add_line(
            [line_pos, pos[1] + 13.0],
            [line_pos, pos[1] + 18.0],
            ig::ImColor32::from_rgb_f32s(1.0, 0.0, 0.0),
        )
        .build();
    }

    id_token.pop(ui);
    changed.then(|| new_frame)
}
